torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -89,12 +355,77 @@ class ChunkDataSet final
   }
 
  private:
+  /// running on worker thread to preload chunk data.
+  void preloader(size_t id) {
+    while (!quit_worker_) {
+      size_t chunk_id;
+      try {
+        auto chunk_sampler_result = chunk_sampler_.next(1);
+        if (chunk_sampler_result.has_value()) {
+          chunk_id = chunk_sampler_result.value()[0];
+        } else {
+          break;
+        }
+        UnwrappedBatchType data = chunk_reader_.read_chunk(chunk_id);
+        if (data.empty()) {
+          if (!ignore_empty_chunk_) {
+            std::string error =
+                "Chunk with index " + std::to_string(chunk_id) + " is empty";
+            throw std::runtime_error(error);
+          } else {
+            // skip adding the current chunk data and move to the next.
+            chunk_buffer_->skip_chunk();
+          }
+        }
+        else {
+          chunk_buffer_->add_chunk_data(chunk_id, std::move(data));
+        }
+      } catch (...) {
+        chunk_buffer_->add_chunk_data(chunk_id, std::current_exception());
+      }
+    }
+  }
+
+  /// Block the current thread until the workers finish execution and exit.
+  void free_workers() {
+    if (!quit_worker_) {
+      quit_worker_ = true;
+      for (auto& worker_thread : preload_threads_) {
+        worker_thread.join();
+      }
+    }
+  }
+
+ private:
+  // chunk reader is responsible for reading chunk data
   ChunkReader chunk_reader_;
-  ChunkSampler chunk_sampler_;
-  ExampleSampler example_sampler_;
-  size_t batch_size_;
-  // Temporary: for API only testing.
-  std::atomic<int> chunk_index_{0};
+
+  // chunk sampler to shuffle different chunks
+  samplers::ThreadSafeSampler<ChunkSamplerType> chunk_sampler_;
+
+  // example sampler to shuffle examples in a specific chunk
+  ExampleSamplerType example_sampler_;
+
+  // chunk data buffer which holds chunk data from preloading thread.
+  std::shared_ptr<BatchDataBuffer<UnwrappedBatchType, ExampleSamplerType>>
+      chunk_buffer_;
+
+  // worker thread pool
+  std::vector<std::thread> preload_threads_;
+
+  // worker thread count
+  size_t preloader_count_ = 0;
```

Comment:
Same comment as before. Should it be initialized on constructor's initialization list?